File: 04_daily_challenge/2021/06-june/week1/max_perf_of_team.py
# https://leetcode.com/explore/challenge/card/june-leetcoding-challenge-2021/603/week-1-june-1st-june-7th/3768/

# Maximum Performance of a Team
# You are given two integers n and k and two integer arrays speed and
# efficiency both of length n. There are n engineers numbered from 1 to n.
# speed[i] and efficiency[i] represent the speed and efficiency of the ith
# engineer respectively.

# Choose at most k different engineers out of the n engineers to form a team
# with the maximum performance.

# The performance of a team is the sum of their engineers' speeds multiplied by
# the minimum efficiency among their engineers.

# Return the maximum performance of this team. Since the answer can be a huge
# number, return it modulo 109 + 7.

# Example 1:
# Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 2
# Output: 60
# Explanation:
# We have the maximum performance of the team by selecting engineer 2 (with
# speed=10 and efficiency=4) and engineer 5 (with speed=5 and efficiency=7).
# That is, performance = (10 + 5) * min(4, 7) = 60.

# Example 2:
# Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 3
# Output: 68
# Explanation:
# This is the same example as the first but k = 3. We can select engineer 1,
# engineer 2 and engineer 5 to get the maximum performance of the team. That
# is, performance = (2 + 10 + 5) * min(5, 4, 7) = 68.

# Example 3:
# Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 4
# Output: 72

# Constraints:
# 1 <= <= k <= n <= 10⁵
# speed.length == n
# efficiency.length == n
# 1 <= speed[i] <= 10⁵
# 1 <= efficiency[i] <= 10⁸

from typing import Callable, List
from termcolor import colored

import heapq


class Solution:
    def maxPerformance(
        self, n: int, speed: List[int], efficiency: List[int], k: int
    ) -> int:
        return self.maxPerformance_greedy_with_priorityq(n, speed, efficiency, k)

    def maxPerformance_greedy_with_priorityq(
        self, n: int, speed: List[int], efficiency: List[int], k: int
    ) -> int:
        candidates = sorted(zip(efficiency, speed), key=lambda t: t[0], reverse=True)
        speed_heap: List[int] = []
        speed_sum, perf = 0, 0
        
        for eff, spd in candidates:
            if len(speed_heap) > k - 1:
                speed_sum -= heapq.heappop(speed_heap)
            heapq.heappush(speed_heap, spd)
            speed_sum += spd
            perf = max(perf, speed_sum * eff)

        return perf % (10 ** 9 + 7)

    def maxPerformance_fastest(
        self, n: int, speed: List[int], efficiency: List[int], k: int
    ) -> int:
        speed_heap: List[int] = []
        speeds, perf = 0, 0
        indexes = list(range(n))
        indexes.sort(key=lambda i: -efficiency[i])

        for i in indexes:
            heapq.heappush(speed_heap, speed[i])
            speeds += speed[i]
            if len(speed_heap) > k:
                lowest_speed = heapq.heappop(speed_heap)
                speeds -= lowest_speed
            perf = max(perf, speeds * efficiency[i])

        return perf % 1000000007


SolutionFunc = Callable[[int, List[int], List[int], int], int]


def test_solution(
    n: int, speed: List[int], efficiency: List[int], k: int, expected: int
) -> None:
    def test_impl(
        func: SolutionFunc,
        n: int,
        speed: List[int],
        efficiency: List[int],
        k: int,
        expected: int,
    ) -> None:
        r = func(n, speed, efficiency, k)
        if r == expected:
            print(
                colored(
                    f"PASSED {func.__name__} => Max performance of chosen {k} of the team with size: {n}, speeds: {speed} efficiencies: {efficiency} is {r}",
                    "green",
                )
            )
        else:
            print(
                colored(
                    f"FAILED {func.__name__} => Max performance of chosen {k} of the team with size: {n}, speeds: {speed} efficiencies: {efficiency} is {r} but expected {expected}",
                    "red",
                )
            )

    sln = Solution()
    test_impl(
        sln.maxPerformance_greedy_with_priorityq, n, speed, efficiency, k, expected
    )
    test_impl(sln.maxPerformance_fastest, n, speed, efficiency, k, expected)


if __name__ == "__main__":
    test_solution(
        n=6, speed=[2, 10, 3, 1, 5, 8], efficiency=[5, 4, 3, 9, 7, 2], k=2, expected=60
    )
    test_solution(
        n=6, speed=[2, 10, 3, 1, 5, 8], efficiency=[5, 4, 3, 9, 7, 2], k=3, expected=68
    )
    test_solution(
        n=6, speed=[2, 10, 3, 1, 5, 8], efficiency=[5, 4, 3, 9, 7, 2], k=4, expected=72
    )

File: 04_daily_challenge/2021/06-june/week1/max_perf_of_team.cpp
// https://leetcode.com/problems/maximum-performance-of-a-team/

// Maximum Performance of a Team

// You are given two integers n and k and two integer arrays speed and
// efficiency both of length n. There are n engineers numbered from 1 to n.
// speed[i] and efficiency[i] represent the speed and efficiency of the ith
// engineer respectively.

// Choose at most k different engineers out of the n engineers to form a team
// with the maximum performance.

// The performance of a team is the sum of their engineers' speeds multiplied
// by the minimum efficiency among their engineers.

// Return the maximum performance of this team. Since the answer can be a huge
// number, return it modulo 109 + 7.

// Example 1:
// Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 2
// Output: 60
// Explanation:
// We have the maximum performance of the team by selecting engineer 2 (with
// speed=10 and efficiency=4) and engineer 5 (with speed=5 and efficiency=7).
// That is, performance = (10 + 5) * min(4, 7) = 60.

// Example 2:
// Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 3
// Output: 68
// Explanation:
// This is the same example as the first but k = 3. We can select engineer 1,
// engineer 2 and engineer 5 to get the maximum performance of the team. That
// is, performance = (2 + 10 + 5) * min(5, 4, 7) = 68.

// Example 3:
// Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 4
// Output: 72

// Constraints:
// 1 <= <= k <= n <= 10⁵
// speed.length == n
// efficiency.length == n
// 1 <= speed[i] <= 10⁵
// 1 <= efficiency[i] <= 10⁸

#include "stdafx.h"
#include <queue>
#include <vector>

using namespace std;
using namespace leetcode::format;

int maxPerformance_greedy_priorityQ(int const n, vector<int> const &speeds,
                                    vector<int> const &efficiencies, int const k)
{
    vector<pair<int, int>> candidates(n);
    for (int i = 0; i < n; i++) {
        candidates[i] = {efficiencies[i], speeds[i]};
    }
    sort(candidates.rbegin(), candidates.rend());

    auto speed_sum = 0, perf = 0;
    priority_queue<int, vector<int>, greater<int>> min_heap;

    for (auto const &[eff, spd] : candidates) {
        min_heap.emplace(spd);
        speed_sum += spd;
        if (min_heap.size() > k) {
            speed_sum -= min_heap.top();
            min_heap.pop();
        }
        perf = max(perf, speed_sum * eff);
    }

    return perf % int(1e9 + 7);
}

void test_solution(int n, vector<int> const &speeds, vector<int> const &efficiencies, int k,
                   int expected)
{
    using SolutionFunc = std::function<int(int, vector<int> const &, vector<int> const &, int)>;

    auto test_impl = [](SolutionFunc func, string_view func_name, int n, vector<int> const &speeds,
                        vector<int> const &efficiencies, int k, int expected) {
        auto const r = func(n, speeds, efficiencies, k);
        if (r == expected) {
            fmt::print(pass_color,
                       "PASSED {} => Max performance of chosen {} of team with size {}, speeds: "
                       "{} and efficiencies: {} is {}.\n",
                       func_name, k, n, to_str(speeds), to_str(efficiencies), r);
        }
        else {
            fmt::print(fail_color,
                       "FAILED {} => Max performance of chosen {} of team with size {}, speeds: "
                       "{} and efficiencies: {} is {} but expected {}.\n",
                       func_name, k, n, to_str(speeds), to_str(efficiencies), r, expected);
        }
    };

    test_impl(maxPerformance_greedy_priorityQ, "maxPerformance_greedy_priorityQ", n, speeds,
              efficiencies, k, expected);
}

int main()
{
    test_solution(6, {2, 10, 3, 1, 5, 8}, {5, 4, 3, 9, 7, 2}, 2, 60);
    test_solution(6, {2, 10, 3, 1, 5, 8}, {5, 4, 3, 9, 7, 2}, 3, 68);
    test_solution(6, {2, 10, 3, 1, 5, 8}, {5, 4, 3, 9, 7, 2}, 4, 72);

    return 0;
}
